ecosystem/sep-0039.md

```diff
@@ -0,0 +1,196 @@
+
+Our exclusive focus in this SEP is Step 2: **storing the asset within the account.**
+
+### Data Format
+We begin by describing the raw data format, then dive into each component:
+
+  - index (2 bytes)
+  - version (1 byte)
+  - metadata length, `n` (varying)
+  - metadata (`n` bytes)
+  - binary data (varying)
+
+For example,
+
+    00|1|26|text/plain;l=13,image/webp|TPwJh>}A"=r@@Y?Fetc...
+
+(Here, the `|` characters are semantic, included just to separate the pieces in the list above; they are not actually in the data itself.)
+
+The following sections break down each component in detail.
+
+#### Index
+Since an account can only have 1000 entries, we only need 10 bits for an index field. We _only_ need indices present so that the data entries are ASCII-sortable in Horizon. In fact, decoders can ignore it entirely if they wish. Thus, the index is expressed as a "base 36" value: 0-9 then a-z.
+
+A valid SEP-39 entry always begins with `00`, indicating that this is the 0th row of data. In fact, every row's key _must_ begin with a two-character, [Base36] index.
+
+---------
+
+**Author's note**: This portion is definitely up for debate. Base36 (well, more accurately, a radix of 36) allows us to shave a byte off of each row (saving up to 1kB total), but we could stick to a 000-999 model (stringified integer) for simplicity. We could also do `aa` through `zz`, because as noted, the indices are _purely_ for sorting. 
+
+Note that supporting a 36-radix is trivial in any language, since we know the value is strictly an integer. For example,
+
+```python
+import string 
+ALPHABET = string.digits + string.ascii_lowercase
+
+def int_to_b36(i: int) -> str:
+    d, r = divmod(i, len(ALPHABET))
+    return ALPHABET[d] + ALPHABET[r]
+
+def b36_to_int(s: str) -> int:
+    return ALPHABET.index(s[0]) * len(ALPHABET) + ALPHABET.index(s[1])
+```
+
+---------
+
+#### Version
+The first entry's index is followed by the version string. For the current version of this SEP, this is just `1`.
+
+#### Metadata
+The **metadata** is a comma-separated list of [Media Type]s alongside any associated parameters.
+
+It is prefixed by a **length**, which is just a stringified integer (that is, a 32-byte metadata will have "32" as its length). When parsing, you should consume characters until hitting a non-digit, since a valid [Media Type] will never start with a digit. Note that row indices are not included in the metadata length.
+
+There is only one standardized parameter:
+
+  * `l`: this defines the *length* (in _raw_ bytes, not encoded data) of this media type within the binary data
+
+You may include others as you see fit.
+
+If there is more than one [Media Type], the length parameter `l` _must_ be set for all but the last type. If there is only one, it may be omitted entirely.
+
+Media types are separated by a comma (`,`) character. For example, this metadata tells us that there text followed by an image within the binary buffer:
+
+    text/plain;l=13,image/png
```

Comment:
Can we also standardize the `k=` field for denoting the `key` field for referencing the asset in question? As this is an NFT SEP it would make sense to have a standard marker mechanic for referencing assets within the binary blob.

I also think the `k=name`, `k=description` and `k=image` should be referenced as standard NFT fields. I'm assuming @FredericRezeau will include these in his off-chain NFT SEP so it would make sense to align some nomenclature here. It shouldn't be required imo but if we want wallets to display name, description and image across the board there will need to be some consensus around standard `k=` fields. If some folks use `n=` or `k=title` things will get messy.

Or should this be a separate SEP? 🤔  @leighmcculloch thoughts?

Reply:
I'm a bit lost. How will an asset be simultaneously text/plain and image/png?

Reply:
So the "asset" is actually a concatenated buffer containing 1 or more assets. So in the Colorglyph case the "asset" will be a name buffer, a description buffer and a png buffer all concatenated together. The metadata details how to split up the buffer back into its individual parts.

Reply:
I'm a bit lost, too, but in a different way: Can you elaborate on what a `key` is? You can already reference a specific chunk because it will follow the metadata order, e.g. the 0th chunk.

Reply:
It's intention is for an NFT naming convention specifically to help wallets parse these sep39 accounts. I'm leaning though that those conventions should be a separate SEP leaning more heavily on @FredericRezeau for input. What values should be included in a sep39 account when being used for an NFT? It will use the `k=` metadata param but what fields are required vs optional when considering the NFT use case?

Reply:
```suggestion
There is only one standardized parameter:

  * `l`: this defines the *length* (in _raw_ bytes, not encoded data) of this media type within the binary data

You may include others as you see fit.

If there is more than one [Media Type], the length parameter `l` _must_ be set for all but the last type. If there is only one, it may be omitted entirely.

Media types are separated by a comma (`,`) character. For example, this metadata tells us that there text followed by an image within the binary buffer:

    text/plain;s=13,image/png
```

Reply:
@Shaptic We still need to resolve https://github.com/stellar/stellar-protocol/pull/1090#discussion_r762303567

Reply:
Ah nvm I see you've added `n=` as a param for the name 👍

Reply:
Just making sure: does `n` cover the `key` case? I don't think we need to add too many standardized fields (though we can add more as needed), because I think the off-chain convention that @FredericRezeau will spearhead will combine itself naturally with this one. An issuer will store metadata in one account (with _all_ of the details you'd expect), then have `nft.uri` (or whatever the standard field will be) point to a Stellar account, which implies said account has the asset in SEP-39 format.

Reply:
> When a wallet encounters an asset issued by an account with sep39 managedata attrs what should hint to it that it's an NFT?

Ah I see your line of reasoning; we avoid the need to depend on another convention. Maybe a `t=nft` as a `type` field? Or it's implied by being an image media type? Or we use the existing `t`itle? Lots of options.

Reply:
My gut says this is a decisions for another SEP around categorizing sep39 usage. i.e. here are metadata fields for NFTs, for real estate, for identity proofs, etc. I also think this may be a good area for the toml standard to come into play, but again should likely be defined in a separate SEP.